pkg/proxy/downstream.go

```diff
@@ -1217,11 +1218,22 @@ func (s *downStream) onUpstreamHeaders(endStream bool) {
 }
 
 func (s *downStream) handleUpstreamStatusCode() {
-	// todo: support config?
 	if s.upstreamRequest != nil && s.upstreamRequest.host != nil {
-		if s.requestInfo.ResponseCode() >= http.InternalServerError {
+		responseCode := s.requestInfo.ResponseCode()
+
+		// 4xx && 5xx
+		if responseCode >= http.BadRequest {
```

Comment:
this could be a break change, I'm not sure if it is a proper choice.

we'd better add some notes for this change, i.e. the description comment of this PR.
people may notice the metrics changed when upgrading to the new mosn.

Reply:
I have modified to make it compatible. As for `UpstreamResponseFailed`, I think it should include all not ok (2xx) conditions, but it wasn't in the past, maybe that's worth discussing. cc @taoyuanyuan

Reply:
nope, 3xx should not be consider as failed.

4xx could be consider as failed, but it's a break change, we could discuss it.

Reply:
> nope, 3xx should not be consider as failed.

My mistake, of source 1xx/2xx/3xx never could be failed. Only 4xx/5xx or status code in configured set could be considered. But I cannot image which condition that 4xx/5xx occered but request is not failed.

Reply:
I'm fine that 4xx to be considered failed, just a bit worried about the break change.

Reply:
okay, we'd better keep backward compatibility.

Reply:
Backward compatibility is no problem, and for those who need it, they can achieve the same effect with simple additions in the Dashboard.

But now failed has exactly the same meaning as server error, maybe we can mark `UpstreamResponse*` as deprecated and remove it in a future version

Reply:
> okay, we'd better keep backward compatibility.

Sorry, my mistake, I realized failed not only occured in L7 but also L4. LGTM